File: native-windows/FitGenerator.Native/Services/AppSettings.cs
using FitGenerator.Native.Map;

namespace FitGenerator.Native.Services;

internal sealed record MapViewportSettings(
    double CenterWgs84Latitude,
    double CenterWgs84Longitude,
    double Zoom)
{
    internal static MapViewportSettings Default => new(
        CenterWgs84Latitude: 39.9042,
        CenterWgs84Longitude: 116.4074,
        Zoom: 9.0);
}

internal sealed record AppSettings
{
    internal const int CurrentSchemaVersion = 1;

    public int SchemaVersion { get; init; } = CurrentSchemaVersion;

    public string ActiveMapProviderId { get; init; } =
        MapProviderDefinition.OpenStreetMapId;

    public IReadOnlyList<MapProviderDefinition> CustomMapProviders { get; init; } =
        Array.Empty<MapProviderDefinition>();

    public MapViewportSettings MapViewport { get; init; } = MapViewportSettings.Default;

    public double PaceSecondsPerKilometer { get; init; } = 360.0;

    public int RestingHeartRate { get; init; } = 60;

    public int MaximumHeartRate { get; init; } = 180;

    public int LapCount { get; init; } = 1;

    public int ExportCount { get; init; } = 1;

    public string? ExportDirectory { get; init; }

    public ulong? PreviewSeed { get; init; }

    internal static AppSettings Default => new();
}

internal sealed record SavedRoute
{
    internal const int CurrentSchemaVersion = 1;

    public int SchemaVersion { get; init; } = CurrentSchemaVersion;

    public IReadOnlyList<GeoCoordinate> Wgs84Points { get; init; } =
        Array.Empty<GeoCoordinate>();

    internal static SavedRoute Empty => new();
}

File: android/app/src/main/java/com/alphaauxiliary/fitgenerator/data/AppSettings.kt
package com.alphaauxiliary.fitgenerator.data

import com.alphaauxiliary.fitgenerator.map.GeoCoordinate
import com.alphaauxiliary.fitgenerator.map.MapProvider

internal data class MapCameraSettings(
    val centerWgs84: GeoCoordinate,
    val zoom: Double,
) {
    companion object {
        val Default = MapCameraSettings(
            centerWgs84 = GeoCoordinate(latitude = 39.9042, longitude = 116.4074),
            zoom = 9.0,
        )
    }
}

internal data class AppSettings(
    val schemaVersion: Int = CURRENT_SCHEMA_VERSION,
    val activeMapProviderId: String = MapProvider.OPEN_STREET_MAP_ID,
    val customMapProviders: List<MapProvider> = emptyList(),
    val mapCamera: MapCameraSettings = MapCameraSettings.Default,
    val paceSecondsPerKilometer: Double = 360.0,
    val restingHeartRate: Int = 60,
    val maximumHeartRate: Int = 180,
    val lapCount: Int = 1,
    val exportCount: Int = 1,
    val previewSeed: ULong? = null,
) {
    override fun toString(): String =
        "AppSettings(schemaVersion=$schemaVersion, activeMapProviderId=$activeMapProviderId, " +
            "customMapProviderCount=${customMapProviders.size}, mapCamera=$mapCamera, " +
            "paceSecondsPerKilometer=$paceSecondsPerKilometer, restingHeartRate=$restingHeartRate, " +
            "maximumHeartRate=$maximumHeartRate, lapCount=$lapCount, exportCount=$exportCount, " +
            "previewSeed=$previewSeed)"

    companion object {
        const val CURRENT_SCHEMA_VERSION = 1
        val Default = AppSettings()
    }
}

internal data class SavedRoute(
    val schemaVersion: Int = CURRENT_SCHEMA_VERSION,
    val wgs84Points: List<GeoCoordinate> = emptyList(),
) {
    companion object {
        const val CURRENT_SCHEMA_VERSION = 1
        val Empty = SavedRoute()
    }
}
